File: sol/solution/2801-2900/2812. Find the Safest Path in a Grid/Solution.ex
defmodule Solution do
  @spec maximum_safeness_factor(grid :: [[integer]]) :: integer
  def maximum_safeness_factor(grid) do
    n = length(grid)
    inf = n * n + 5

    dist = :array.new(n * n, default: inf)
    queue = :queue.new()

    {dist, queue} =
      Enum.with_index(grid)
      |> Enum.reduce({dist, queue}, fn {row, r}, {d, q} ->
        Enum.with_index(row)
        |> Enum.reduce({d, q}, fn {cell, c}, {d2, q2} ->
          if cell == 1 do
            idx = r * n + c
            {:array.set(idx, 0, d2), :queue.in({r, c}, q2)}
          else
            {d2, q2}
          end
        end)
      end)

    dirs = [{1, 0}, {-1, 0}, {0, 1}, {0, -1}]
    dist = bfs(queue, dist, n, dirs)

    start_safe = :array.get(0, dist)
    end_safe = :array.get(n * n - 1, dist)
    hi = min(start_safe, end_safe)

    binary_search(0, hi, dist, n, dirs)
  end

  defp bfs(queue, dist, n, dirs) do
    case :queue.out(queue) do
      {:empty, _} ->
        dist

      {{:value, {r, c}}, q} ->
        cur = :array.get(r * n + c, dist)

        {dist2, q2} =
          Enum.reduce(dirs, {dist, q}, fn {dr, dc}, {d, qu} ->
            nr = r + dr
            nc = c + dc

            if nr >= 0 and nr < n and nc >= 0 and nc < n do
              idx = nr * n + nc

              if :array.get(idx, d) > cur + 1 do
                {:array.set(idx, cur + 1, d), :queue.in({nr, nc}, qu)}
              else
                {d, qu}
              end
            else
              {d, qu}
            end
          end)

        bfs(q2, dist2, n, dirs)
    end
  end

  defp binary_search(lo, hi, dist, n, dirs) when lo < hi do
    mid = div(lo + hi + 1, 2)

    if can_reach?(mid, dist, n, dirs) do
      binary_search(mid, hi, dist, n, dirs)
    else
      binary_search(lo, mid - 1, dist, n, dirs)
    end
  end

  defp binary_search(lo, _hi, _dist, _n, _dirs), do: lo

  defp can_reach?(limit, dist, n, dirs) do
    if :array.get(0, dist) < limit do
      false
    else
      visited = :array.new(n * n, default: false)
      visited = :array.set(0, true, visited)
      queue = :queue.in({0, 0}, :queue.new())

      dfs(queue, visited, limit, dist, n, dirs)
    end
  end

  defp dfs(queue, visited, limit, dist, n, dirs) do
    case :queue.out(queue) do
      {:empty, _} ->
        false

      {{:value, {r, c}}, q} ->
        if r == n - 1 and c == n - 1 do
          true
        else
          {visited2, q2} =
            Enum.reduce(dirs, {visited, q}, fn {dr, dc}, {vis, qu} ->
              nr = r + dr
              nc = c + dc

              if nr >= 0 and nr < n and nc >= 0 and nc < n do
                idx = nr * n + nc

                if not :array.get(idx, vis) and :array.get(idx, dist) >= limit do
                  {:array.set(idx, true, vis), :queue.in({nr, nc}, qu)}
                else
                  {vis, qu}
                end
              else
                {vis, qu}
              end
            end)

          dfs(q2, visited2, limit, dist, n, dirs)
        end
    end
  end
end

File: sol/solution/2801-2900/2812. Find the Safest Path in a Grid/Solution.c
typedef struct
{
    int x, y;
} Node;

int maximumSafenessFactor(int **grid, int gridSize, int *gridColSize)
{
    int n = gridSize;
    const int INF = 1000000000;

    int *distMem = (int *)malloc(n * n * sizeof(int));
    int **dist = (int **)malloc(n * sizeof(int *));
    for (int i = 0; i < n; i++)
    {
        dist[i] = distMem + i * n;
    }

    Node *queue = (Node *)malloc(n * n * sizeof(Node));
    int front = 0, back = 0;

    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            if (grid[i][j] == 1)
            {
                dist[i][j] = 0;
                queue[back++] = (Node){i, j};
            }
            else
            {
                dist[i][j] = INF;
            }
        }
    }

    int dx[4] = {1, -1, 0, 0};
    int dy[4] = {0, 0, 1, -1};

    while (front < back)
    {
        Node cur = queue[front++];

        for (int k = 0; k < 4; k++)
        {
            int nx = cur.x + dx[k];
            int ny = cur.y + dy[k];

            if (nx < 0 || ny < 0 || nx >= n || ny >= n)
                continue;

            if (dist[nx][ny] > dist[cur.x][cur.y] + 1)
            {
                dist[nx][ny] = dist[cur.x][cur.y] + 1;
                queue[back++] = (Node){nx, ny};
            }
        }
    }

    if (dist[0][0] == 0 || dist[n - 1][n - 1] == 0)
    {
        free(queue);
        free(dist);
        free(distMem);
        return 0;
    }

    char *visited = (char *)malloc(n * n);

    int left = 0, right = 2 * n;

    while (left < right)
    {
        int mid = (left + right + 1) / 2;

        if (dist[0][0] < mid || dist[n - 1][n - 1] < mid)
        {
            right = mid - 1;
            continue;
        }

        memset(visited, 0, n * n);

        front = back = 0;
        queue[back++] = (Node){0, 0};
        visited[0] = 1;

        int found = 0;

        while (front < back)
        {
            Node cur = queue[front++];

            if (cur.x == n - 1 && cur.y == n - 1)
            {
                found = 1;
                break;
            }

            for (int k = 0; k < 4; k++)
            {
                int nx = cur.x + dx[k];
                int ny = cur.y + dy[k];

                if (nx < 0 || ny < 0 || nx >= n || ny >= n)
                    continue;

                int id = nx * n + ny;

                if (visited[id])
                    continue;

                if (dist[nx][ny] < mid)
                    continue;

                visited[id] = 1;
                queue[back++] = (Node){nx, ny};
            }
        }

        if (found)
            left = mid;
        else
            right = mid - 1;
    }

    free(visited);
    free(queue);
    free(dist);
    free(distMem);

    return left;
}